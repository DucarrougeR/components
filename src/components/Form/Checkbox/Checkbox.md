```js
<Checkbox id="show-filter-bar" name="themeShowFilterBars" label="Show Filter Bars" />
<Checkbox id="show-title-bar" name="themeTitleBars" label="Show Title Bars" />
```
