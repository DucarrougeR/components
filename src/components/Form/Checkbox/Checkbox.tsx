// Typescript component boilerplate for generating meaningful declaration files.
import * as React from 'react'
import styled, { StyledComponentClass } from '../../../styled_components'
export { StyledComponentClass }
import { ThemeInterface } from '../../../themes'
export { ThemeInterface }
// End Typescript component boilerplate

import { FormGroup } from '../FormGroup/FormGroup'
import { Input, InputTypes } from '../Input/Input'
import { Label } from '../Label/Label'

export interface CheckboxProps {
  id: string
  label: string
  name: string
  placeholder?: string
  required?: boolean
}

/**
 *   Checkbox are a way for a user to make a range of selections
 */
const CheckboxGenerator: React.SFC<CheckboxProps> = ({
  id,
  label,
  name,
  required,
  ...args
}) => {
  // // This prevents our props from being passed directly to the underlying tags, which ultimately
  // // would cause some warnings. Ideally we would define the return type for this function, and
  // // Typescript would warn us when passing props that are invalid.
  // //
  // // See https://reactjs.org/warnings/unknown-prop.html
  // const props = Object.assign({}, args)
  // delete props.id
  // delete props.label
  // delete props.name
  // delete props.required

  return (
    <FormGroup>
      <Label htmlFor={id}>{label}</Label>
      <input
        id={id}
        type="checkbox"
        name={name}
        required={required}
        {...args}
      />
    </FormGroup>
  )
}

export const Checkbox = styled<CheckboxProps>(CheckboxGenerator)``
