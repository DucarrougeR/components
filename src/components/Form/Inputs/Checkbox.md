```js noeditor
const StatusAndResources = require('../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Form/Inputs/Checkbox.tsx"
  feedbackTitle="Checkbox Component Feedback" />
```

### Default Checkbox

By default, the `<Checkbox />` component renders an `<input type="checkbox" />` tag.

```js
  <Checkbox />
```

### Name and ID

A name and ID can be specified in the `<Checkbox />` component. Names are important if the checkbox is used in the context of a form, in which case a name is required for the value of the checkbox to be captured.

```js
  <Checkbox name="someName" id="someId" />
```

### Checked Property

The `<Checkbox />` component accepts the `checked` property, allowing the user to force the checkbox as always checked or never checked.

```js
  <Checkbox checked={true} />
  <Checkbox checked={false} />
```

### Disabled Property

Use the disable property to gray out the checkbox, making it unclickable.

```js
  <Checkbox disabled />
  <Checkbox disabled checked />

```

### Alt Property

Should the checkbox fail to render, `alt` text will be substituted in its place.
