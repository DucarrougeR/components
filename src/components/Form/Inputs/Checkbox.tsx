import tag from 'clean-tag'
import * as React from 'react'
import styled from '../../../styled_components'
import { reset } from '../../../styles/reset'
import { InputProps } from './InputProps'

export interface CheckboxProps extends InputProps {
  checked?: boolean
}

const InternalCheckbox: React.SFC<CheckboxProps> = ({ ...props }) => {
  return <tag.input type="checkbox" {...props} />
}

export const Checkbox = styled<CheckboxProps>(InternalCheckbox)`
  ${reset};
`
