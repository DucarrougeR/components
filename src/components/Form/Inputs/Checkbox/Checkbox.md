```js noeditor
const StatusAndResources = require('../../../../../styleguide_components/StatusAndResources').StatusAndResources;

<StatusAndResources
  status="experimental"
  figmaURL="https://www.figma.com/file/2MG6DoSjk4IaLnjjFCnKFf/Forms"
  githubURL="https://github.com/looker/lens/blob/master/src/components/Form/Inputs/Checkbox.tsx"
  feedbackTitle="Checkbox Component Feedback" />
```

### Default Checkbox

The `<Checkbox />` component renders a single checkbox element on the page, with no accompanying label. It is generally used to construct higher-order components like the `<FieldCheckbox />`. If you are building a form, you probably want to use `<FieldCheckbox />` instead as it provides label support.

```js
<Checkbox />
```

### Name and ID

A name and ID can be specified in the `<Checkbox />` component. Names are important if the checkbox is used in the context of a form, in which case a name is required for the value of the checkbox to be captured.

```js
<Checkbox name="someName" id="someId" />
```

### Checked Property

The `<Checkbox />` component accepts the `checked` property, allowing the user to force the checkbox as always checked or never checked.

```js
<Checkbox checked={true} />
<Checkbox checked={false} />
```

### Disabled Property

Use the disable property to gray out the checkbox, making it unclickable.

```js
<Checkbox disabled />
<Checkbox disabled checked />
```
