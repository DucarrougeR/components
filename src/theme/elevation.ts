export type ElevationLayers = '1' | '2' | '3' | '4' | '5' | '6' | '7' | 'max'

export type Shadows = { [K in ElevationLayers]?: string }

export const shadows: Shadows = {
  1: '0 1px 2px rgba(0, 0, 0, 0.1), 0px 1px 1px rgba(0, 0, 0, 0.05)',
  2: '0 3px 4px rgba(0, 0, 0, 0.11), 0 1px 2px rgba(0, 0, 0, 0.04)',
  3: '0 6px 10px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.04)',
  4: '0 9px 18px rgba(0, 0, 0, 0.14), 0 1px 2px rgba(0, 0, 0, 0.04),  0 10px 8px 4px rgba(0, 0, 0, 0.01)',
  5: '0 14px 20px rgba(0, 0, 0,.15), 0 1px 1px rgba(0, 0, 0, 0.04),  0 14px 10px 8px rgba(0, 0, 0, 0.02)',
  6: '0 22px 42px 0px rgba(0, 0, 0,.18),  0 1px 1px rgba(0, 0, 0, 0.05), 0 16px 20px 10px rgba(0, 0, 0, 0.025)',
}
