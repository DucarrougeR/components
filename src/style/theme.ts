import { rem } from 'polished'
import { InterpolationValue } from 'styled-components'
import {
  SizeLarge,
  SizeMedium,
  SizeNone,
  SizeSmall,
  SizeXLarge,
  SizeXSmall,
  SizeXXLarge,
  SizeXXXLarge,
} from '../types'
import { Easings, easings } from './easings'
import { shadows, Shadows } from './elevation'
import { fontFaces, FontFaces } from './font_faces'
import { FontRamp, fontSizes, lineHeights } from './font_sizes'
import { fontWeights } from './font_weights'
import { palette, Palette } from './palette'
import { SemanticColors, semanticColors } from './semantic_colors'
import { transitions, Transitions } from './transitions'

export * from './easings'
export * from './elevation'
export * from './font_sizes'
export * from './palette'
export * from './semantic_colors'
export * from './transitions'

export type SpacingSizes =
  | SizeNone
  | SizeLarge
  | SizeMedium
  | SizeSmall
  | SizeXLarge
  | SizeXSmall
  | SizeXXLarge
  | SizeXXXLarge

export interface Theme {
  breakpoints: string[]
  easings: Easings
  fontFaces: FontFaces
  fontSizes: FontRamp
  fontWeights: Record<string, number>
  lineHeights: FontRamp
  palette: Palette
  /**
   * A function that can be overridden to return different reset css properties
   * or null to remove all resets. Most base elements in Lens implement the reset.
   */
  reset: () => InterpolationValue
  semanticColors: SemanticColors
  shadows: Shadows
  space: Record<SpacingSizes, string>
  transitions: Transitions
}

export const theme: Theme = {
  breakpoints: ['30rem', '48rem', '64rem', '75rem', '90rem'],
  easings,
  fontFaces,
  fontSizes,
  fontWeights,
  lineHeights,
  palette,
  reset: () => {
    return {
      border: 0,
      font: 'inherit',
      fontSize: '100%',
      margin: 0,
      padding: 0,
      verticalAlign: 'baseline',
    }
  },
  semanticColors,
  shadows,
  space: {
    // tslint:disable:object-literal-sort-keys
    none: 'none',
    xsmall: rem(4),
    small: rem(8),
    medium: rem(16),
    large: rem(20),
    xlarge: rem(32),
    xxlarge: rem(40),
    xxxlarge: rem(60),
    // tslint:enable:object-literal-sort-keys
  },
  transitions,
}
