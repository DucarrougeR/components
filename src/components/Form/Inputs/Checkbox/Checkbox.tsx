import * as React from 'react'
import { styled } from '../../../../style'
import { Box, BoxProps } from '../../../Box'
import { InputProps } from '../InputProps'

export interface CheckboxProps extends BoxProps<HTMLInputElement>, InputProps {
  /**
   * Determines if the checkbox is checked or not.
   */
  checked?: boolean
}

const InternalCheckbox: React.FC<CheckboxProps> = ({
  validationType,
  ...props
}) => {
  const type = { type: 'checkbox' }
  return <Box is="input" {...props} {...type} />
}

export const Checkbox = styled<CheckboxProps>(InternalCheckbox)``
