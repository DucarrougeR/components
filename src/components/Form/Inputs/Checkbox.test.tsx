import { mount } from 'enzyme'
import 'jest-styled-components'
import * as React from 'react'
import { create } from 'react-test-renderer'
import theme from '../../../theme'
import { Checkbox } from './Checkbox'

test('Checkbox default', () => {
  const component = create(<Checkbox theme={theme} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Checkbox checked set to true', () => {
  const component = create(<Checkbox theme={theme} checked={true} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Checkbox checked set to false', () => {
  const component = create(<Checkbox theme={theme} checked={false} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Checkbox with name and id', () => {
  const component = create(<Checkbox theme={theme} name="Chuck" id="Chucky" />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Checkbox should accept disabled', () => {
  const component = create(<Checkbox theme={theme} disabled />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Checkbox with aria-describedby', () => {
  const component = create(
    <Checkbox theme={theme} aria-describedby="some-id" />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Should trigger onChange handler', () => {
  let counter = 0
  const handleChange = () => counter++

  const wrapper = mount(<Checkbox theme={theme} onChange={handleChange} />)

  wrapper.find('input').simulate('change', { target: { value: '' } })
  expect(counter).toEqual(1)
})
