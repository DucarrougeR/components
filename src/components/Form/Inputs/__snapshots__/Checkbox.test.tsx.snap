// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox checked set to false 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  checked={false}
  className="c0"
  type="checkbox"
/>
`;

exports[`Checkbox checked set to true 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  checked={true}
  className="c0"
  type="checkbox"
/>
`;

exports[`Checkbox default 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  className="c0"
  type="checkbox"
/>
`;

exports[`Checkbox should accept disabled 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  className="c0"
  disabled={true}
  type="checkbox"
/>
`;

exports[`Checkbox with aria-describedby 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  aria-describedby="some-id"
  className="c0"
  type="checkbox"
/>
`;

exports[`Checkbox with name and id 1`] = `
.c0 {
  margin: 0;
  padding: 0;
  border: 0;
  font: inherit;
  font-size: 100%;
  vertical-align: baseline;
}

<input
  className="c0"
  id="Chucky"
  name="Chuck"
  type="checkbox"
/>
`;
